src/Chapter3.hs

```diff
@@ -480,6 +513,42 @@ After defining the city, implement the following functions:
    and at least 10 living __people__ inside in all houses of the city totally.
 -}
 
+data Castle = Castle(String) | CastleWithWalls(String) | None deriving (Eq, Show)
+
+data Building = Church | Library deriving Show
+
+data PeopleAmount = One | Two | Three | Four deriving Show
+
+data House = House {
+  peopleAmount :: PeopleAmount
+} deriving Show
+
+data City = City {
+  castle :: Castle,
+  building :: Building,
+  houses :: [House]
+} deriving Show
+
+buildCastle :: City -> String -> City
+buildCastle city castleName = city{castle = Castle(castleName)}
+
+buildHouse :: City -> House -> City
+buildHouse city@City{houses} house = city{houses = house:houses}
+
+
+buildWalls :: City -> City
+buildWalls city@City{castle,houses} = case castle of
+  Castle castleName -> if pplCount > 10 then city{castle = CastleWithWalls(castleName)} else city
+  _ -> city
+  where
+    pplCount = foldl (\acc house' -> acc + (pplInHouse . peopleAmount $ house')) (0 :: Int) houses
```

Comment:
Nice usage of standard functions! If you want to use `foldl`, then it's always better to use `foldl'` from the `Data.List` module. Ordinary `foldl` leaks memory, unfortunately :potable_water: 

Alternatively, you can use a combination of `map` and `sum` functions from `Prelude`. But your solution is already correct, these are just some notes :slightly_smiling_face: